topdown/crypto.go

```diff
@@ -378,42 +417,69 @@ func getX509CertsFromPem(pemBlocks []byte) ([]*x509.Certificate, error) {
 	return x509.ParseCertificates(decodedCerts)
 }
 
-func getRSAPrivateKeyFromString(key string) (interface{}, error) {
-	// if the input is PEM handle that
+func getPrivateKeyFromString(key string) (interface{}, error) {
 	if strings.HasPrefix(key, "-----BEGIN") {
-		return getRSAPrivateKeyFromPEM([]byte(key))
+		return getPrivateKeyFromPem(key)
 	}
 
-	// assume input is base64 if not PEM
 	b64, err := base64.StdEncoding.DecodeString(key)
 	if err != nil {
 		return nil, err
 	}
-
-	return getRSAPrivateKeyFromPEM(b64)
+	return getPrivateKeyFromPem(string(b64))
 }
 
-func getRSAPrivateKeyFromPEM(pemBlocks []byte) (interface{}, error) {
-
-	// decode the pem into the Block struct
-	p, _ := pem.Decode(pemBlocks)
-	if p == nil {
-		return nil, fmt.Errorf("failed to parse PEM block containing the key")
+func getPrivateKeyFromPem(pemBlocks string) (interface{}, error) {
+
+	var validPrivateKeys []crypto.PrivateKey
+
+	if len(pemBlocks) != 0 {
+
+		bs := []byte(pemBlocks)
+		for len(bs) > 0 {
+			inputLen := len(bs)
+			var block *pem.Block
+			block, bs = pem.Decode(bs)
+			if block == nil && len(bs) == 0 {
+				break
+			}
+
+			if inputLen == len(bs) {
```

Comment:
This case, as I understand it, only happens when the end of the input is not a valid PEM block. I think it's worth explaining what we're handling here and covering with a test case like I have in https://go.dev/play/p/2XqZn1wbgC2

Reply:
I added the test in https://github.com/open-policy-agent/opa/pull/6084/commits/28f166de3d0571a23aabde8bb7e36b176ba71132 and description in https://github.com/open-policy-agent/opa/pull/6084/commits/4a9b6e46bd27d35feeb14fe9661cde41d7ab62bb